taichi/backends/opengl/opengl_api.h

```diff
@@ -11,17 +11,21 @@ TLANG_NAMESPACE_BEGIN
 
 namespace opengl {
 
-struct GLProgram;
 void initialize_opengl();
 bool is_opengl_api_available();
 void create_glsl_root_buffer(size_t size);
 void begin_glsl_kernels(const std::vector<IOV> &iov);
-void launch_glsl_kernel(GLProgram *program, int num_groups);
 void end_glsl_kernels(const std::vector<IOV> &iov);
-GLProgram *compile_glsl_program(std::string source);
 int opengl_get_threads_per_group();
 extern bool opengl_has_GL_NV_shader_atomic_float;
 
+struct GLProgram;
+struct GLProgramIFace {
```

Comment:
Does `IFace` mean `Interface`? I understand networking people use this abbreviation, but in C++ `IFace` doesn't seem to be a common abbreviation. Would `Interface` be better? It's not much longer anyway.

Reply:
I suddenly realized it's not that `interface` in object-oriented languages, like the java one. It's just a OO-ish OpenGL API, so I changed it into `CompiledGLSL` now, as a base of `CompiledKernel`.